extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheResultInterceptor.java

```diff
@@ -19,11 +24,63 @@
+
         CaffeineCache cache = cacheRepository.getCache(binding.cacheName());
         Object key = getCacheKey(cache, binding.cacheKeyParameterPositions(), context.getParameters());
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debugf("Loading entry with key [%s] from cache [%s]", key, cache.getName());
         }
-        return cache.get(key, () -> context.proceed(), binding.lockTimeout());
+
+        if (binding.lockTimeout() <= 0) {
+            CompletableFuture<Object> cacheValue = cache.get(key, (k, executor) -> {
+                return getValueLoader(context, executor);
+            });
+            return cacheValue.get();
+        } else {
+
+            // The lock timeout logic starts here.
+
+            /*
+             * If the current key is not already associated with a value in the Caffeine cache, there's no way to know if the
+             * current thread or another one started the missing value computation. The following variable will be used to
+             * determine whether or not a timeout should be triggered during the computation depending on which thread started
+             * it.
+             */
+            boolean[] isCurrentThreadComputation = { false };
+
+            CompletableFuture<Object> cacheValue = cache.get(key, (k, executor) -> {
+                isCurrentThreadComputation[0] = true;
+                return getValueLoader(context, executor);
+            });
+
+            if (isCurrentThreadComputation[0]) {
+                // The value is missing and its computation was started from the current thread.
+                // We'll wait for the result no matter how long it takes.
+                return cacheValue.get();
+            } else {
+                // The value is either already present in the cache or missing and its computation was started from another thread.
+                // We want to retrieve it from the cache within the lock timeout delay.
+                try {
+                    return cacheValue.get(binding.lockTimeout(), TimeUnit.MILLISECONDS);
+                } catch (TimeoutException e) {
+                    // Timeout triggered! We don't want to wait any longer for the value computation and we'll simply invoke the
+                    // cached method and return its result without caching it.
+                    // TODO: Add statistics here to monitor the timeout.
+                    return context.proceed();
+                }
+            }
+        }
+    }
+
+    private CompletableFuture<Object> getValueLoader(InvocationContext context, Executor executor) {
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                return context.proceed();
+            } catch (RuntimeException e) {
+                throw e;
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }, executor);
```

Comment:
General comment on this class: we try to avoid lambdas in runtime code.

Reply:
Thanks for reviewing this!

I already knew about lambdas being the enemy in runtime code but there are only functional interfaces instances in this class (`BiFunction` for lines 35 and 51 and `Supplier` for line 76) which were already present before the PR (with a few differences). Are those nasty lambdas which should be eliminated as well?

Reply:
You can do it in a subsequent PR if you prefer but I would get rid of all the lambdas in this class.

Reply:
I removed all lambdas from the extension runtime code, it should be good now.